### PR TITLE
Use colour-coding for user disambiguation

### PR DESCRIPTION
**Pull request by a-andreyev**
Contributes to issue #513 
I like this HSL-based appoarch so tried to help (implemented similar for r0kk3rz Sailfish OS version), but feel to reject/correct :)

Checked with different KDE color schemes:

Imgur

### CODE REVIEW

**Comment by KitsuneRal** [comment]
Thanks for the start! And yes, HSL-based calculation is certainly the way to go. We have a few more tasks ahead though :)
- [ ] Calculate colour in the library (https://github.com/QMatrixClient/libqmatrixclient/issues/296)
- [ ] Colour users in the user list as well
- [ ] Provide a switch to use colour disambiguation instead of full-name disambiguation

**Comment by a-andreyev** [comment]
current results screenshot:

Imgur

**Comment by KitsuneRal** [comment]
Now that we have the lib PR, it's time to rebase this PR on the newest `master` :)

File: client/models/userlistmodel.cpp [hunk 1]
```diff
@@ -108,6 +110,15 @@ QVariant UserListModel::data(const QModelIndex& index, int role) const
         return tooltip;
     }
 
+    if (role == Qt::TextColorRole)
```
Comment (KitsuneRal) [inline client/models/userlistmodel.cpp]: I'd rather we use a custom role; `TextColorRole` is about the "main" text colour and is too easy to confuse with the message colour.

File: client/qml/TimelineItem.qml [hunk 2]
```diff
@@ -46,14 +46,35 @@ Item {
...
+        if (str) {
+            return Qt.hsla(author.hueF,
+                           (1-defaultPalette.window.hslSaturation),
+                           /* contrast but not too heavy: */
+                           (-0.7*defaultPalette.window.hslLightness + 0.9),
+                           textColor.a)
+        }
+        return textColor
+    }
+    readonly property color fancyColor: stringToColor(authorName)
```
Comment (KitsuneRal) [inline client/qml/TimelineItem.qml]: You don't need all this remake here now that the library gives you everything necessary - just get that custom role I mentioned above.

File: client/qml/TimelineItem.qml [hunk 3]
```diff
@@ -156,7 +177,7 @@ Item {
                     actionEvent ? Text.AlignRight : Text.AlignLeft
                 elide: Text.ElideRight
 
-                color: textColor
+                color: fancyColor
```
Comment (KitsuneRal) [inline client/qml/TimelineItem.qml]: ...and insert that role name here.

File: client/qml/TimelineItem.qml [hunk 4]
```diff
@@ -240,7 +261,7 @@ Item {
                     // FIXME: The text is clumsy and slows down creation
                     text: (actionEvent && !xchatStyle ?
                            ("<a href='#mention' style='text-decoration:none;color:\"" +
-                                    defaultPalette.text + "\"'><b>" +
+                                    fancyColor + "\"'><b>" +
```
Comment (KitsuneRal) [inline client/qml/TimelineItem.qml]: ...and here.

File: client/models/userlistmodel.cpp [hunk 5]
```diff
@@ -22,6 +22,8 @@
 #include <QElapsedTimer>
 #include <QtCore/QDebug>
 #include <QtGui/QPixmap>
+#include <QPalette>
+#include <QApplication>
```
Comment (KitsuneRal) [inline client/models/userlistmodel.cpp]: Qt module names, please.

**Comment by a-andreyev** [comment]
rebased and testing, getting black color for now (checking what versrion I've compiled)

File: client/models/userlistmodel.cpp (hunk 1 above)
Comment (a-andreyev) [reply client/models/userlistmodel.cpp]: not sure how to connect `TextColorRole` with custom role for a  `QTableView` at `UserListDock` (widget-based ui part), not changed for now

**Comment by a-andreyev** [comment]
working now, but I have a boilerplate code for final color calculation from hue: for `UserListDock` widget c++ part and `TimelineItem` qml part. Not sure yet where's better place to provide shared solution :thinking:

File: client/models/userlistmodel.cpp (hunk 1 above)
Comment (KitsuneRal) [reply client/models/userlistmodel.cpp]: Ah, we don't have any custom roles there yet. Ok, keep it as is now.

**Comment by KitsuneRal** [comment]
At the moment, all CI pipelines are breaking because you seem to have forgotten to add the newest lib to your commit. In order to do that, you have to make sure the contents under `lib/` subdirectory are at the right commit (basically - the newest commit in `master`) and then commit the subdirectory itself.

File: client/models/userlistmodel.cpp (hunk 1 above)
Comment (KitsuneRal) [reply client/models/userlistmodel.cpp]: Ok, there's still one thing to fix here: according to Qt docs, `TextColorRole` is obsolete, and `ForegroundRole` should be used instead.

**Comment by KitsuneRal** [comment]
> working now, but I have a boilerplate code for final color calculation from hue: for `UserListDock` widget c++ part and `TimelineItem` qml part. Not sure yet where's better place to provide shared solution 🤔

Well, a fancy way of doing it would be to introduce a Quaternion-specific user class (`QuaternionUser`) instead of the generic `QMatrixClient::User` (`Connection` allows you to customise the class it uses to create user objects) and provide `QuaternionUser::color()` that also backs a `Q_PROPERTY` so that it could be referred to from QML. If that sounds a bit too much for you, just leave the boilerplate - we'll get back to it sometime later.

**Comment by a-andreyev** [comment]
> If that sounds a bit too much for you, just leave the boilerplate - we'll get back to it sometime later.

Probably a bit too much but sounds interesting :slightly_smiling_face: Agree to leave it for now and save the task as an issue for example

**Pull request merged**